packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts

```diff
@@ -0,0 +1,132 @@
+    ) {
+        const noopStorage = { getItem: () => null, setItem: () => {} };
+        if (localStorageImpl !== undefined) {
+            this.localStorage = localStorageImpl;
+        } else {
+            try {
+                // localStorage is not defined in Node environment so this throws
+                this.localStorage = localStorage ?? noopStorage;
+            } catch (error) {
+                this.localStorage = noopStorage;
+            }
+        }
+
+        if (this.localStorage === noopStorage) {
+            // This means the Blackout Period logic will not work.
+            this.mc.logger.sendTelemetryEvent({ eventName: "SweepReadyUsageDetectionHandlerNoOpStorage" });
+        }
+    }
+
+    /**
+      * If SweepReady Usage Detection is enabled, close the main container.
+      * If the "Blackout Period" is set, don't close the container more than once in that period.
+      *
+      * Once Sweep is fully implemented, this will be removed since the objects will be gone
+      * and errors will arise elsewhere in the runtime
+     */
+    public usageDetectedInMainContainer(errorProps: ITelemetryProperties) {
+        if (!sweepReadyUsageDetectionSetting.read(this.mc.config).mainContainer) {
+            return;
+        }
+
+        const pastClosuresMap: Record<string, { lastCloseTime: number; } | undefined> = (() => {
+            try {
+                const rawValue = this.localStorage.getItem(closuresStorageKey);
+                const parsedValue = rawValue === null ? {} : JSON.parse(rawValue);
+                return typeof parsedValue === "object"
+                    ? parsedValue as Record<string, { lastCloseTime: number; } | undefined>
+                    : {};
+            } catch (e) {
+                return {};
+            }
+        })();
+
+        const lastCloseTime = pastClosuresMap[this.uniqueContainerKey]?.lastCloseTime;
+        const blackoutPeriodDays = this.mc.config.getNumber(blackoutPeriodDaysKey);
+
+        const error = new SweepReadyUsageError(
+            "SweepReady object used in Non-Summarizer Client",
+            { errorDetails: JSON.stringify({ ...errorProps, lastCloseTime, blackoutPeriodDays }) },
+        );
+
+        // Should close if...
+        const shouldClose = lastCloseTime === undefined // ...We've not closed before, or...
+            || blackoutPeriodDays === undefined // ...there's no blackout duration set, or...
+            || Date.now() > lastCloseTime + blackoutPeriodDays * oneDayMs; // ...we've passed the blackout period
+
+        if (shouldClose) {
+            // Update closures map in localStorage before closing
+            // Note there is a race condition between different tabs updating localStorage and overwriting
+            // each others' updates. If so, some tab will crash again. Just reload one at a time to get unstuck
+            pastClosuresMap[this.uniqueContainerKey] = { lastCloseTime: Date.now() };
```

Comment:
Is this key supposed to be unique per document or per container? Currently, you are using `ContainerContext::id` which is the document id.

Reply:
Per document.  Needs to prevent the next session from closing.

Reply:
Basically if 10 different clients open the document, they will all close once if I understand this correctly.

Reply:
This bug will be very hard to reproduce on the consumer end

Reply:
Tyler - You are right that if we go this route, we'll lose a lot of leverage with users in terms of getting good bug reports.  When I first reach out to partners, I'm not planning to mention this capability.  But it's there, so if the conversation starts to go towards "We have to disable SweepReady Usage Detection, the user impact is too high", then we will bring this out.

In other words, it gives us a dial on user pain so we can tune it to the org's tolerance for 💩 user experience in DF

Reply:
Also, not sure you said it quite right - the pastClosuresMap is in localStorage in that browser (or Teams or whatever).  So every user gets the crash once per "browser" per X days.  Still pretty annoying IMO.

Reply:
True, I did not try to get that specific